### PR TITLE
SCD219 Allow FilterPeaks to select peaks by bank name (master)

### LINKED ISSUE #30726: SCD219 Allow peaks in a peaks workspace to be filtered by banks using FilterPeaks algorithm
The current `FilterPeaks` algorithm can only filter peaks by numerical values.
We would like to extend its capability to be able to select peaks from an input peakworkspace by bank name (string).

The original Gitlab user story is [SCD219](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/219)

### PR DESCRIPTION
**Pull request by KedoKudo**
**Description of work.**

This PR extend the capability of `FilterPeaks` such that users can select peaks based on bank name, in addition to filtering by property values.

- Related Gitlab task [SCD215](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/215)
- This PR is #30730 into `master`

**To test:**
Use [data.tar.gz](https://github.com/mantidproject/mantid/files/6007283/data.tar.gz) and the script below for testing inside workbench
```python
from mantid.simpleapi import *

pws = LoadIsawPeaks('data/Natrolite_runs_133752_133812.peaks', )

FilterPeaks(
    InputWorkspace="pws",
    OutputWorkspace="pws_intensity",
    FIlterVariable='Intensity',
    FilterValue=100,
    Operator='>',
)

FilterPeaks(
    InputWorkspace="pws",
    OutputWorkspace="pws_bank7",
    BankName='bank7',
    Criterion='=',
)

FilterPeaks(
    InputWorkspace="pws",
    OutputWorkspace="pws_intensity_bank7",
    FIlterVariable='Intensity',
    FilterValue=10,
    Operator='>',
    BankName='bank7',
    Criterion='=',
)
```

Fixes #30726. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by peterfpeterson — DISMISSED** [review]
Clear code to add new functionality. Additional test verify it works.

File: Framework/Crystal/inc/MantidCrystal/FilterPeaks.h [hunk 1]
```diff
@@ -71,6 +72,33 @@ class MANTID_CRYSTAL_DLL FilterPeaks : public API::Algorithm {
...
+   * @tparam Comparator
+   * @param inputWS :: the input peaks workspace containing peaks to be filtered
+   * @param filteredWS :: the output peaks workspace which will contain a subset
+   * of the inputWS
+   * @param filterFunction :: function extracting the value to filter on from
+   * the peak object
+   * @param filterValue :: the string value to check
+   */
+  template <typename Comparator>
+  void filterPeaksStr(const Mantid::API::IPeaksWorkspace &inputWS,
```
Comment (jmborr) [inline Framework/Crystal/inc/MantidCrystal/FilterPeaks.h:89]: `Comparator` is meant to be a function, not a type. To me, the following function signature seems more appropriate:
```C++
template <typename Comparator>
void filterPeaksStr(const Mantid::API::IPeaksWorkspace &inputWS,
                    Mantid::API::IPeaksWorkspace &filteredWS,
                    const FilterFunctionStr &filterFunction,
                    const std::string filterValue,
                    Comparator operatorFunc);
```
Later you could do:
```C++
filterPeaksStr(*inputWS, *selectedWS, filterFunction, bankname, std::equal_to<std::string>());
```
The advantage of this is that you could pass custom Comparator functions.
See [here for a nice working example](https://godbolt.org/z/3pwX4G).

**Review by jmborr — CHANGES_REQUESTED** [review]

File: Framework/Crystal/inc/MantidCrystal/FilterPeaks.h [hunk 2]
```diff
@@ -71,6 +72,33 @@ class MANTID_CRYSTAL_DLL FilterPeaks : public API::Algorithm {
         filteredWS.addPeak(currentPeak);
     }
   }
+
+  /**
+   * @brief Select peaks from the input peak workspace by checking a string
+   *        value
+   *
+   * @tparam Comparator
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply Framework/Crystal/inc/MantidCrystal/FilterPeaks.h:80]: `tparam` stands for template parameter.

File: Framework/Crystal/src/FilterPeaks.cpp [hunk 3]
```diff
@@ -96,31 +120,53 @@ void FilterPeaks::exec() {
...
+    PeaksWorkspace_sptr selectedWS = filteredWS->clone();
+
+    if (criterion == "=")
+      filterPeaksStr<std::equal_to<std::string>>(*inputWS, *selectedWS,
+                                                 filterFunction, bankname);
+    else if (criterion == "!=")
+      filterPeaksStr<std::not_equal_to<std::string>>(*inputWS, *selectedWS,
+                                                     filterFunction, bankname);
+    else
+      throw std::invalid_argument("Unsupport operator " + criterion +
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply Framework/Crystal/src/FilterPeaks.cpp:137]: typo fixed

File: Framework/Crystal/inc/MantidCrystal/FilterPeaks.h (hunk 1 above)
Comment (KedoKudo) [reply Framework/Crystal/inc/MantidCrystal/FilterPeaks.h:89]: The current style in use
`filterPeaks<std::less<double>>(*inputWS, *filteredWS, filterFunction,
                                     filterValue);` is already decided by the previous developer, and I am only here to extending its supported type from double to string.
I would prefer not include style change as it is beyond the scope of this PR.

Comment (jmborr) [reply Framework/Crystal/inc/MantidCrystal/FilterPeaks.h:89]: I guess that's fair 😃

**Review by jmborr — APPROVED** [review]

**Pull request merged**